obsplus/datasets/dataset.py

```diff
@@ -1,65 +1,91 @@
 from distutils.dir_util import copy_tree
 from functools import lru_cache
 from pathlib import Path
 from types import MappingProxyType as MapProxy
-from typing import Union, Optional, Tuple
+from typing import Union, Optional, Tuple, TypeVar
 from warnings import warn
 
-import obspy.clients.fdsn
 import pkg_resources
 from obspy.clients.fdsn import Client
 
 import obsplus
-from obsplus import WaveBank, EventBank, copy_dataset
+from obsplus import copy_dataset
 from obsplus.constants import DATA_TYPES
-from obsplus.datasets.utils import _create_opsdata
-from obsplus.events.utils import get_event_client
 from obsplus.exceptions import (
     FileHashChangedError,
     MissingDataFileError,
     DataVersionError,
 )
-from obsplus.stations.utils import get_station_client
-from obsplus.utils import md5_directory
-from obsplus.waveforms.utils import get_waveform_client
+from obsplus.interfaces import WaveformClient, EventClient, StationClient
+from obsplus.utils.dataset import _create_opsdata
+from obsplus.utils.events import get_event_client
+from obsplus.utils.misc import hash_directory
+from obsplus.utils.stations import get_station_client
+from obsplus.utils.waveforms import get_waveform_client
+
+DataSetType = TypeVar("DataSetType", bound="DataSet")
 
 
 class DataSet(abc.ABC):
     """
-    Class for downloading and serving datasets.
+    Abstract Base Class for downloading and serving datasets.
 
-    By default the data will be downloaded to obsplus' datasets module
-    but this can be changed using the base_path argument. All data will
-    be saved in base_path / name.
+    This is not intended to be used directly, but rather through subclassing.
 
     Parameters
     ----------
     base_path
-        The path to which the dataset will be saved as base_path / name.
+        The path to which the dataset will be saved.
 
     Attributes
     ----------
+    data_path
+        The path containing the data. By default it is base_path / name.
     source_path
-        The path to the directory containing the source of DataSet code.
+        The path which contains the original files included in the dataset
+        before download. By default this is found in the same directory as
+        the dataset's code (.py) file in a folder with the same name as the
+        dataset.
+
+
+    Notes
+    -----
+        Importantly, each dataset references *two* directories, the source_path
+        and data_path. The source_path contains all data included which the
+        dataset and should not be altered. The data_path has a copy of
+        everything in the source_path, plus the files created during the
+        downloading process.
+
+        The base_path (the parent of data_path) is resolved for each
+        dataset using the following priorities:
+
+            1. The `base_path` provided to `Dataset`'s __init__ method.
+            2. .data_path.txt file stored in the data source
+            3. An environmental name OPSDATA_PATH
```

Comment:
> I would think the files should go wherever the user wants them

Yes, but I want to give the user fine-grained control over this as well. 1) allows the user to specify, per dataset instance, where it is stored. 2) The .data_path.txt allows different datasets in different places (eg in case some are large and require their own drive). 3) simply overwrite the default for *all* datasets.

Reply:
Gotcha. That makes sense. I guess I interpreted 1 and 2 as things that the person who created the datasets defines.